lib/src/machine.rs

```diff
@@ -0,0 +1,379 @@
+    /// stack reference is invalid or the stack is at capacity, an error is
+    /// returned.
+    fn push(
+        &mut self,
+        env: &Environment,
+        stack_id: StackIdentifier,
+        value: LangValue,
+    ) -> Fallible<()> {
+        let stack = self.get_stack(stack_id)?;
+
+        // If the stack is capacity, make sure we're not over it
+        if let Some(max_stack_size) = env.max_stack_size {
+            if stack.len() == max_stack_size {
+                return Err(RuntimeError::StackOverflow(stack_id).into());
+            }
+        }
+
+        stack.push(value);
+        Ok(())
+    }
+
+    /// Pops an element off the given stack. If the pop is successful, the
+    /// popped value is returned. If the stack doesn't exist or is empty, an
+    /// error is returned.
+    fn pop(&mut self, stack_id: StackIdentifier) -> Fallible<LangValue> {
+        let stack = self.get_stack(stack_id)?;
+
+        if let Some(val) = stack.pop() {
+            Ok(val)
+        } else {
+            Err(RuntimeError::EmptyStack(stack_id).into())
+        }
+    }
+}
+
+/// The current state of a machine. This encompasses the entire state of a
+/// program that is currently being executed.
+///
+/// The fields are public for read-only purposes. They should never be mutated.
+/// Initialized from an [Environment](Environment), which controls the input
+/// and stack parameters.
+#[derive(Debug, PartialEq)]
+pub struct MachineState {
+    /// The current input buffer. This can be popped from as the program is
+    /// executed. This will be initialized as the reverse of the input from the
+    /// environment, so that elements at the beginning can be popped off first.
+    /// Values can never be added to the input, only popped off.
+    pub input: Vec<LangValue>,
+    /// The current output buffer. This can be pushed into, but never popped
+    /// out of.
+    pub output: Vec<LangValue>,
+    /// The mutable variable that the programmer has direct control over.
+    pub workspace: LangValue,
+    /// The series of stacks that act as the programs RAM. The number of stacks
+    /// and their capacity is determined by the initialization environment.
+    pub stacks: Stacks,
+}
+
+impl MachineState {
+    fn new(env: &Environment) -> Self {
+        let mut input = env.input.clone();
+        input.reverse(); // Reverse the vec so we can pop off the values in order
+        Self {
+            stacks: Stacks::new(&env),
+            input,
+            output: Vec::new(),
+            workspace: 0,
+        }
+    }
+}
+
+/// A recursive structure to track the progression of a program. A program
+/// counter consist of a flat list of instructions that need to be run, plus
+/// an optional child `ProgramCounter`. If the child is present, the child's
+/// instructions will always be executed before any of the parent's. Once all
+/// of a program counter's instructions have been executed, it's considered
+/// exhausted.
```

Comment:
Very clean way to handle blocks. I approve with 4.9 belching burps out of 5

Reply:
oh fuqqqqq ye bud